include/matx/transforms/cub.h

```diff
@@ -1949,6 +2076,127 @@ void sort_impl(OutputTensor &a_out, const InputOperator &a,
 #endif
 }
 
+
+
+/**
+ * Sort rows of a tensor
+ *
+ * Sort rows of a tensor using a radix sort. Currently supported types are
+ * float, double, ints, and long ints (both signed and unsigned). For a 1D
+ * tensor, a linear sort is performed. For 2D and above each row of the inner
+ * dimensions are batched and sorted separately. There is currently a
+ * restriction that the tensor must have contiguous data in both rows and
+ * columns, but this restriction may be removed in the future.
+ *
+ * @note Temporary memory is used during the sorting process, and about 2N will
+ * be allocated, where N is the length of the tensor.
+ *
+ * @tparam T1
+ *   Type of data to sort
+ * @tparam RANK
+ *   Rank of tensor
+ * @param idx_out
+ *   Indices of sorted tensor
+ * @param a
+ *   Input operator
+ * @param dir
+ *   Direction to sort (either SORT_DIR_ASC or SORT_DIR_DESC)
+ * @param exec
+ *   CUDA executor
+ */
+template <typename OutputTensor, typename InputOperator>
+void argsort_impl(OutputTensor &idx_out, const InputOperator &a,
+          const SortDirection_t dir,
+          cudaExecutor exec = 0)
+{
+#ifdef __CUDACC__
+  MATX_NVTX_START("", matx::MATX_NVTX_LOG_API)
+
+  static constexpr int RANK = OutputTensor::Rank();
+
+  using a_type = typename InputOperator::value_type;
+  a_type *a_ptr = nullptr;
+  a_type *a_out_ptr = nullptr;
+  index_t *idx_in_ptr = nullptr;
+  detail::tensor_impl_t<a_type, InputOperator::Rank()> tmp_a;
+  detail::tensor_impl_t<a_type, InputOperator::Rank()> tmp_a_out;
+  detail::tensor_impl_t<index_t, InputOperator::Rank()> tmp_idx_in;
+    
+  // sorting currently requires a contiguous tensor view, so allocate a temporary
+  // tensor to copy the input if necessary.
+  bool use_a = false;
+  if constexpr (is_tensor_view_v<InputOperator>) {
+    if (a.IsContiguous()) {
+      make_tensor(tmp_a, a.Data(), a.Shape());
```

Comment:
More of a comment for the future, but I think a merge sort supports input iterators and not just tensors. This could eliminate the use of the temporary tensors.

Reply:
It does, but I think it still mutates the keys iterator (i.e. the input tensor).

I wasn't sure exactly which sorting method to use for these. Should we make it an option?

Reply:
It's probably worth asking the CUB people. I know they are faster at different things and I don't recall which.

Reply:
It seems like the general advice is:
- for 1D, use `DeviceRadixSort`
- for ND, use `DeviceSegmentedSort`

Reply:
What about MergeSort?

Reply:
The general advice was to only use MergeSort if you need a custom operator.

In our case, we would still need allocate a temporary as merge sort mutates the input values.